### ISSUE TITLE
Compiling a 15M Parameter Agent from Scratch on Apple Silicon — What Actually Worked

### ISSUE THREAD

**Issue by chan4lk** [chan4lk/blog#8]
We took an OKR management agent that routes queries to 5 API tools and trained a tiny transformer to replace Claude API calls. The model runs locally in 10MB. Getting there was harder than the theory predicted.

## Goal

Build a task-specific micro-model that handles workflow routing and tool-call generation for an OKR agent, trained from scratch on synthetic data, running entirely on an M3 Pro Mac — no cloud GPU, no API costs at inference time.

The approach is called **Agent-Specific Model Synthesis (ASMS)**: use a large LLM as a compiler (once, offline) to generate training data, then train a tiny model sized to the agent's actual decision complexity.

## The Agent

The OKR agent manages goals using the Keyflow MCP API with 5 tools, 20 operations, and 6 workflows:

- `goal_to_okr` — translate business goals into structured OKRs
- `view_okrs` — list objectives and key results
- `check_in` — update progress scores
- `reports` — generate health checks and summaries
- `onboard` — create starter OKRs for new hires
- `align` — cascade objectives across org levels

Effective decision space: ~500 distinct paths. The theory says a 15M parameter model should handle this. Theory was right — eventually.

## The Pipeline

```
Role Spec (YAML) → Synthetic Corpus (Claude Sonnet agents)
    → BPE Tokenizer (SentencePiece) → MLX Transformer (from scratch)
    → INT4 Quantization → OpenAI-compatible Server + Chat UI
```

Everything runs locally via Apple MLX on Metal GPU. Training data generated by spawning parallel Sonnet agents within Claude Code sessions — $0 in API fees.

### Attempt 1: Small model, small corpus (5.7M params, 299 examples)

The model output nothing but UNK tokens. Teacher-forced accuracy was 71.7%, but autoregressive generation collapsed completely. We couldn't tell if anything was being learned.

### Attempt 2: More data, same model (5.7M params, 5,759 examples)

Accuracy barely moved — 80.7%. The model learned workflow routing (5/10 correct) but couldn't produce valid JSON. We hit a wall: **the model was too small**, not undertrained.

### Attempt 3: Bigger model, same data (15M params, 5,759 examples)

This was the breakthrough. Same training data, just more model capacity:

| Metric | 5.7M model | 15M model | Change |
|--------|-----------|-----------|--------|
| Workflow routing | 5/10 | 8/10 | +60% |
| Valid JSON tool calls | 2/10 | 5/10 | +150% |

The 15M model produced correct, parseable Keyflow MCP calls:

```json
{"tool": "objective", "action": "list", "params": {"cycleId": "cyc_q2_2026", "ownerId": "usr_107"}}
{"tool": "key_result", "action": "check_in", "params": {"keyResultId": "kr_102", "value": 1}}
{"tool": "report", "action": "health_check", "params": {"cycleId": "cyc_q2_2026"}}
```

### Attempt 4: Bigger corpus, bigger model (15M params, 8,859 examples)

This made things worse. Routing dropped from 8/10 to 3/10. We'd rebuilt the tokenizer for the larger corpus, which shifted all vocabulary IDs and destroyed the learned embeddings.

### 1. Model capacity beats data volume

Doubling the data from 1K to 5K barely moved accuracy (+9pp). Tripling the model parameters on the same data jumped it +60% on routing. Size the model to the decision space first, then worry about corpus.

### 2. Never rebuild the tokenizer

This was the most expensive lesson. SentencePiece assigns vocabulary IDs during training. Rebuild the tokenizer and every embedding the model learned becomes meaningless. Freeze the tokenizer after the first training run. Add data? Retokenize with the same `.model` file.

### 3. Overfitting is fast and silent

With 15M parameters and 5K examples, the best checkpoint was at epoch 3. Training loss kept falling to 0.05 but validation loss climbed steadily. The final model was always worse than the early one. Use 8-10 epochs max with aggressive early stopping.

### 4. Temperature 0.0 is non-negotiable

Temperature 0.1 caused the model to get stuck generating space tokens in a loop — decoded as UNK characters. Greedy decoding (temperature=0.0) was the only thing that produced structured output. This cost us hours of debugging.

### 5. The effort is real

Even for a "simple" 500-decision-space agent:
- 8,859 training examples across 9 generation rounds
- 5 architecture iterations
- 6 training runs (48s to 8 hours each)
- Multiple tokenizer/overfitting/temperature debugging sessions

The ASMS paper makes it sound like "just compile your agent." The compilation works, but the compiler needs calibration.

## Final Architecture

```
Model:       6-layer decoder-only transformer
Params:      15M (10MB quantized INT4)
Hidden:      384, Heads: 6, FFN: 768
Vocab:       6,000 BPE tokens
Context:     512 tokens
Framework:   Apple MLX (Metal GPU)
Training:    58 min on M3 Pro
Corpus:      5,759 synthetic examples
Best result: 80% workflow routing, 50% valid JSON tool calls
```

## How to Try It

The model is on HuggingFace: [chan4lk/okr-micro-asms](https://huggingface.co/chan4lk/okr-micro-asms)

```bash
git clone https://github.com/chan4lk/timm
cd timm && uv sync
uv run deploy/server.py model/checkpoints/best
# Open http://localhost:8800
```

The chat UI connects directly to Keyflow MCP — toggle "Live MCP" to execute real tool calls from the micro-model's predictions.

## What's Next

The 80% routing accuracy validates the ASMS thesis: a 15M model can replace a trillion-parameter LLM for bounded tool-use decisions. The gap to 95% likely needs:

1. **Constrained JSON decoding** — enforce grammar rules during generation
2. **Frozen-tokenizer incremental training** — add corpus without resetting embeddings
3. **Two-stage architecture** — separate classifier head for routing + decoder for JSON

The paper (with empirical results): [Agent-Specific Model Synthesis v0.2](https://github.com/chan4lk/timm/blob/main/docs/2026-03-31-agent-specific-model-synthesis.md)
